### ISSUE TITLE
ara_record fails to decrypt SSL on https endpoint GET /api/v1/results

### ISSUE THREAD

**Issue by djangoxv** [ansible-community/ara#101]
The ara_record module seems to have an issue instantiating the results api

Ansible 2.9.1, python 3.6.1, ara and ara[server] 1.3.0
I run the API server in django using nginx proxy+uwsgi on Debian bare metal

I am using a simple playbook to record the local branch
```- name: Record local git workspace branch
  hosts: 127.0.0.1
  connection: local
  become: no
  vars:
    ara_playbook_name: ara-record
  tasks:
    - name: Lookup local git branch
      command: "git rev-parse --abbrev-ref HEAD"
      register: git_branch
      changed_when: false
      check_mode: no

    - name: Record git branch in ARA
      ara_record:
        key: "active_git_branch"
        value: "{{ git_branch.stdout }}"
      ignore_errors: "{{ ansible_check_mode }}"
```

This *works*, tested in 1.2.0 and 1.30 - so thank you - we still can record our git branch

But there is a PYTHON_WARNING thrown on SSL decryption:
```
TASK [Record git branch in ARA] ****************************************************************************************************
task path: /ansible_2.9.1/playbooks/ara-record.yml:19
changed: [127.0.0.1] => {
    "changed": true,
    "created": "2019-12-07T14:28:50.050584-08:00",
    "key": "active_git_branch",
    "msg": "Record created or updated in ARA",
    "playbook_id": 100,
    "type": "text",
    "updated": "2019-12-07T14:28:50.050629-08:00",
    "value": "master"
}
[WARNING]: Failure using method (v2_runner_on_ok) in callback plugin (<ansible.plugins.callback.ara_default.CallbackModule object
at 0x7fa429e70ef0>): HTTPSConnectionPool(host='verified.site.com', port=443): Max retries exceeded with url: /api/v1/results
(Caused by SSLError(SSLError(1, '[SSL: DECRYPTION_FAILED_OR_BAD_RECORD_MAC] decryption failed or bad record mac (_ssl.c:2178)'),))

Callback Exception:
  File "/ansible_2.9.1/venv/lib/python3.6/site-packages/ansible/executor/task_queue_manager.py", line 323, in send_callback
    method(*new_args, **kwargs)
   File "/ansible_2.9.1/venv/lib/python3.6/site-packages/ara/plugins/callback/ara_default.py", line 261, in v2_runner_on_ok
    self._load_result(result, "ok", **kwargs)
   File "/ansible_2.9.1/venv/lib/python3.6/site-packages/ara/plugins/callback/ara_default.py", line 373, in _load_result
    ignore_errors=kwargs.get("ignore_errors", False) or False,
   File "/ansible_2.9.1/venv/lib/python3.6/site-packages/ara/clients/http.py", line 115, in post
    return self._request("post", endpoint, **kwargs)
   File "/ansible_2.9.1/venv/lib/python3.6/site-packages/ara/clients/http.py", line 93, in _request
    response = func(url, **kwargs)
   File "/ansible_2.9.1/venv/lib/python3.6/site-packages/ara/clients/http.py", line 69, in post
    return self._request("post", url, data=json.dumps(payload))
   File "/ansible_2.9.1/venv/lib/python3.6/site-packages/ara/clients/http.py", line 57, in _request
    return self.http.request(method, self.endpoint + url, timeout=self.timeout, **payload)
   File "/ansible_2.9.1/venv/lib/python3.6/site-packages/requests/sessions.py", line 533, in request
    resp = self.send(prep, **send_kwargs)
   File "/ansible_2.9.1/venv/lib/python3.6/site-packages/requests/sessions.py", line 646, in send
    r = adapter.send(request, **kwargs)
   File "/ansible_2.9.1/venv/lib/python3.6/site-packages/requests/adapters.py", line 514, in send
    raise SSLError(e, request=request)
```

I assume something is lost here, but whatever is lost by not grabbing `/api/v1/results`

Thanks!
P.S. Keep on doing ARA up... glad to see so much movement since 1.x

**Comment by dmsimard** [comment]
Haven't looked at this issue yet but wanted to point out that I've been seeing it as well:
```
TASK [smoke-tests : Record data with no type] **********************************
task path: /home/zuul/src/opendev.org/recordsansible/ara/tests/integration/roles/smoke-tests/tasks/ara-ops.yaml:30
changed: [localhost] => {
    "changed": true,
    "created": "2020-01-30T00:15:25.989098Z",
    "key": "notype",
    "msg": "Record created or updated in ARA",
    "playbook_id": 6,
    "type": "text",
    "updated": "2020-01-30T00:15:25.989138Z",
    "value": "text"
}
[WARNING]: Failure using method (v2_runner_on_ok) in callback plugin
(<ansible.plugins.callback.ara_default.CallbackModule object at
0x7fac33c8f910>): HTTPSConnectionPool(host='api.demo.recordsansible.org',
port=443): Max retries exceeded with url: /api/v1/results (Caused by
SSLError(SSLError(1, '[SSL: DECRYPTION_FAILED_OR_BAD_RECORD_MAC] decryption
failed or bad record mac (_ssl.c:2508)')))
Callback Exception: 
  File "/home/zuul/.ara-tests/virtualenv/lib64/python3.7/site-packages/ansible/executor/task_queue_manager.py", line 324, in send_callback
    method(*new_args, **kwargs)
   File "/home/zuul/.ara-tests/virtualenv/lib/python3.7/site-packages/ara/plugins/callback/ara_default.py", line 283, in v2_runner_on_ok
    self._load_result(result, "ok", **kwargs)
   File "/home/zuul/.ara-tests/virtualenv/lib/python3.7/site-packages/ara/plugins/callback/ara_default.py", line 395, in _load_result
    ignore_errors=kwargs.get("ignore_errors", False) or False,
   File "/home/zuul/.ara-tests/virtualenv/lib64/python3.7/site-packages/ara/clients/http.py", line 115, in post
    return self._request("post", endpoint, **kwargs)
   File "/home/zuul/.ara-tests/virtualenv/lib64/python3.7/site-packages/ara/clients/http.py", line 93, in _request
    response = func(url, **kwargs)
   File "/home/zuul/.ara-tests/virtualenv/lib64/python3.7/site-packages/ara/clients/http.py", line 69, in post
    return self._request("post", url, data=json.dumps(payload))
   File "/home/zuul/.ara-tests/virtualenv/lib64/python3.7/site-packages/ara/clients/http.py", line 57, in _request
    return self.http.request(method, self.endpoint + url, timeout=self.timeout, **payload)
   File "/home/zuul/.ara-tests/virtualenv/lib64/python3.7/site-packages/requests/sessions.py", line 533, in request
    resp = self.send(prep, **send_kwargs)
   File "/home/zuul/.ara-tests/virtualenv/lib64/python3.7/site-packages/requests/sessions.py", line 646, in send
    r = adapter.send(request, **kwargs)
   File "/home/zuul/.ara-tests/virtualenv/lib64/python3.7/site-packages/requests/adapters.py", line 514, in send
    raise SSLError(e, request=request)
```

**Comment by dmsimard** [comment]
I've investigated this a bit but haven't found the root cause yet.
It looks like perhaps there could be something wrong with our usage of weakref for re-using the client instance ?

tl;dr:
1) Client sends POST requests to create the playbook, the play, the task without issues
2) Client sends POST request to create a record from ara_record without issues
3) After the ara_record task, the callback gets a SSLError exception when trying to POST the result of the task

The relevant order of events:
1) callback creates a new http client instance: https://github.com/ansible-community/ara/blob/b1e450b9bd77609c741e4bd667b2fd4f9c5e8e25/ara/plugins/callback/ara_default.py#L187-L194
2) ``v2_playbook_on_start`` runs fine: https://github.com/ansible-community/ara/blob/b1e450b9bd77609c741e4bd667b2fd4f9c5e8e25/ara/plugins/callback/ara_default.py#L196
3) ``v2_playbook_on_play_start`` runs fine: https://github.com/ansible-community/ara/blob/b1e450b9bd77609c741e4bd667b2fd4f9c5e8e25/ara/plugins/callback/ara_default.py#L221
4) ``v2_playbook_on_task_start`` runs fine: https://github.com/ansible-community/ara/blob/b1e450b9bd77609c741e4bd667b2fd4f9c5e8e25/ara/plugins/callback/ara_default.py#L261
5) the task started is an ara_record task
6) ara_record recovers the active client instance through weakref: https://github.com/ansible-community/ara/blob/b1e450b9bd77609c741e4bd667b2fd4f9c5e8e25/ara/plugins/action/ara_record.py#L143
7) ara_record uses client.post to create the new key/value: https://github.com/ansible-community/ara/blob/b1e450b9bd77609c741e4bd667b2fd4f9c5e8e25/ara/plugins/action/ara_record.py#L150
8) the task completes successfully, invoking ``v2_runner_on_ok``
9) ``v2_runner_on_ok`` raises an SSLError exception when doing a post for the result of the task: https://github.com/ansible-community/ara/blob/b1e450b9bd77609c741e4bd667b2fd4f9c5e8e25/ara/plugins/callback/ara_default.py#L402
10) The remainder of the playbook runs fine and is recorded properly (aside from the results of ara_record tasks)

**Comment by dmsimard** [comment]
Yeah, the issue has something to do with passing the instance weakref back and forth.

If I revert the [patch](https://github.com/ansible-community/ara/commit/b632aae7d82a25e3a60d834e5ccee938eddc9fec#diff-e0a8a9a7b6f45ca04a4d939a1f547329) that introduced the weakref, I am unable to reproduce the issue.

**Comment by dmsimard** [comment]
I haven't found a definitive fix for this yet and it didn't warrant blocking the release of 1.4, we can fix it in an upcoming dot release.

**Comment by dmsimard** [comment]
I haven't revisited this issue in a long time but perhaps this PR could fix it: https://github.com/ansible-community/ara/pull/274

**Comment by djangoxv** [comment]
FYI - when I updated to 1.5.8 recently I noticed that setting the callback_threads=1 in ansible.cfg solved this issue
I suspect when you bump to Django 3 you will find it resolved

**Comment by hille721** [comment]
Issue still exists in 1.6.1. With setting `callback_threads` no exception is comming anymore, but still the results are not stored in Ara. With new tasks page this is good visible:

![image](https://user-images.githubusercontent.com/38468371/225063125-bdd8069c-7f94-478b-8153-c7f16e305a63.png)
